tlatools/org.lamport.tlatools/src/tla2sany/parser/JListStack.java

```diff
@@ -0,0 +1,137 @@
+package tla2sany.parser;
+
+import java.util.Stack;
+
+/**
+ * Handling vertically-aligned conjunction & disjunction lists (called JLists
+ * for short) is one of the most difficult parts of parsing TLA+. This class
+ * keeps track of the current JList parsing context by storing a stack of the
+ * nested JLists, their start columns, and their types.
+ */
+public class JListStack {
+
+  /**
+   * Enumeration of JList types.
+   */
+  private enum JListType {
+    /**
+     * Conjunction lists, which use the /\ bullet symbol.
+     */
+    CONJUNCTION,
+
+    /**
+     * Disjunction lists, which use the \/ bullet symbol.
+     */
+    DISJUNCTION
+  }
+
+  /**
+   * A class holding information about a JList.
+   */
+  private class JListInfo {
+
+    /**
+     * The type of the JList.
+     */
+    public final JListType Type;
+
+    /**
+     * The column on which the JList starts. This uses codepoints, not bytes,
+     * and is the number of codepoints before the start of the bullet symbol.
+     */
+    public final int Column;
+
+    /**
+     * Constructs a new instance of the JListInfo class.
+     *
+     * @param type The JList type.
+     * @param column The JList start column.
+     */
+    public JListInfo(JListType type, int column) {
+      this.Type = type;
+      this.Column = column;
+    }
+  }
+
+  /**
+   * A stack tracking the nested JLists for the current parse context.
+   */
+  private final Stack<JListInfo> stack = new Stack<JListInfo>();
+
+  /**
+   * Constructs a new instance of the JListStack class with an empty stack.
+   */
+  public JListStack() { }
+
+  /**
+   * Determines whether the given token kind is a JList start token.
+   *
+   * @param kind The token kind.
+   * @return Whether the given token is a JList start token.
+   */
+  private static boolean isJListKind(int kind) {
+    return TLAplusParserConstants.AND == kind
+      || TLAplusParserConstants.OR == kind;
+  }
+
+  /**
+   * Resolves the given parser constant kind to either a conjunction or
+   * disjunction JList.
+   *
+   * @param kind The parser constant kind.
```

Comment:
Here (and perhaps in other methods) I would add that `kind` is `one of the {@link TLAplusParserConstants}`. (As a reader of the TLA+ tools source code, I often struggle to figure out which of the zillion pseudo-enums a given `int` represents.)